alica_engine/src/engine/PlanBase.cpp

```diff
@@ -22,407 +22,372 @@
+                                _deepestNode->getAssignment()->getRobotStateMapping()->getRobotsInState(
+                                        _deepestNode->getActiveState()).end(),
+                                back_inserter(_statusMessage->robotIDsWithMe));
+
+                    }
+                    else
+                    {
+                        _statusMessage->currentState = "NONE";
+                    }
+                    auto tmpRole = _ra->getOwnRole();
+                    if(tmpRole)
+                    {
+                        _statusMessage->currentRole = _ra->getOwnRole()->getName();
+                    }
+                    else
+                    {
+                        _statusMessage->currentRole = "No Role";
+                    }
+                    _ae->getCommunicator()->sendAlicaEngineInfo(*_statusMessage);
+                    _lastSentStatusTime = _alicaClock->now();
+                }
+            }
+
+            _log->iterationEnds(_rootNode);
+
+            _ae->iterationComplete();
+
+            AlicaTime availTime;
+
+            now = _alicaClock->now();
+            if (_loopTime > (now - beginTime))
+            {
+                //TODO: FIXME: the division by 1000 is brittle and should not be needed
+                //replace AlicaTime with a proper time class.
+                availTime = (AlicaTime)((_loopTime - (now - beginTime)) / 1000UL);
+            }
+            else
+            {
+                availTime = 0;
+            }
+            //TODO: FIXME: Need to wait for a signal from an fpEvent here.
+
+            if (_fpEvents.size() > 0)
+            {
+                //lock for fpEvents
+                {
+                    lock_guard<mutex> lock(_lomutex);
+                    while (_running && availTime > 1000 && _fpEvents.size() > 0)
+                    {
+                        shared_ptr<RunningPlan> rp = _fpEvents.front();
+                        _fpEvents.pop();
+
+                        if (rp->isActive())
+                        {
+                            bool first = true;
+                            while (rp != nullptr)
+                            {
+                                cout << "TICK FPEVENT " << endl;
+                                PlanChange change = _ruleBook.visit(rp);
+                                cout << "AFTER TICK FPEVENT " << endl;
+                                if (!first && change == PlanChange::NoChange)
+                                {
+                                    break;
+                                }
+                                rp = rp->getParent().lock();
+                                first = false;
+                            }
+                        }
+                        now = _alicaClock->now();
+                        if (_loopTime > (now - beginTime))
+                        {
+                            availTime = (long)((_loopTime - (now - beginTime)) / 1000UL);
```

Comment:
Change to AlicaTime here, too.

Reply:
Yes, will come in one of the next PRs.
Actually, we'll make a new AlicaTime class while at it..